nvflare/edge/widgets/device_runner.py

```diff
@@ -0,0 +1,212 @@
+from nvflare.apis.fl_constant import FLContextKey, ReservedKey, ReturnCode
+from nvflare.apis.fl_context import FLContext
+from nvflare.edge.constants import EdgeApiStatus, EdgeEventType, Status
+from nvflare.edge.simulated_device import DeviceFactory, DeviceState, SimulatedDevice
+from nvflare.edge.web.models.result_report import ResultReport
+from nvflare.edge.web.models.result_response import ResultResponse
+from nvflare.edge.web.models.task_request import TaskRequest
+from nvflare.edge.web.models.task_response import TaskResponse
+from nvflare.fuel.f3.message import Message as CellMessage
+from nvflare.fuel.utils.validation_utils import check_number_range, check_positive_int, check_str
+from nvflare.security.logging import secure_format_exception
+from nvflare.widgets.widget import Widget
+
+
+class DeviceRunner(Widget):
+
+    def __init__(self, device_factory_id: str, num_devices=100, num_workers=10):
+        Widget.__init__(self)
+
+        check_str("device_factory_id", device_factory_id)
+        check_positive_int("num_devices", num_devices)
+        check_positive_int("num_workers", num_workers)
+        check_number_range("num_workers", num_workers, 1, num_devices)
+        check_number_range("num_workers", num_workers, 1, 100)
+
+        self.device_factory_id = device_factory_id
+        self.num_devices = num_devices
+        self.device_factory = None
+        self.done = False
+        self.job_id = None
+        self.engine = None
+        self.devices = []
+        self.worker_pool = ThreadPoolExecutor(num_workers)
+
+        self.register_event_handler(EventType.ABOUT_TO_START_RUN, self._dr_about_to_start)
+        self.register_event_handler(EventType.START_RUN, self._dr_start_run)
+        self.register_event_handler(EventType.END_RUN, self._dr_end_run)
+
+    def _dr_about_to_start(self, event_type: str, fl_ctx: FLContext):
+        is_leaf = fl_ctx.get_prop(ReservedKey.IS_LEAF)
+        if not is_leaf:
+            # devices are only for leaf nodes
+            return
+
+        self.log_info(fl_ctx, "device runner about to start ...")
+        engine = fl_ctx.get_engine()
+        factory = engine.get_component(self.device_factory_id)
+        if not isinstance(factory, DeviceFactory):
+            self.system_panic(
+                f"component {self.device_factory_id} must be DeviceFactory but got {type(factory)}",
+                fl_ctx,
+            )
+            return
+
+        self.engine = engine
+        self.device_factory = factory
+        self.job_id = fl_ctx.get_job_id()
+        self.log_info(fl_ctx, f"got device factory {type(factory)} for job {self.job_id}")
+
+    def _dr_start_run(self, event_type: str, fl_ctx: FLContext):
+        is_leaf = fl_ctx.get_prop(ReservedKey.IS_LEAF)
+        if not is_leaf:
+            # devices are only for leaf nodes
+            return
+
+        for _ in range(self.num_devices):
+            try:
+                device = self.device_factory.make_device()
```

Comment:
Good for initial experiments, thinking about millions device, if initiallized in one-shot (for-loop in 156), not sure whether it will become a burden (interval will be very small if num_devices is huge). We may further simulate more complicated mode for devices "dynamically added" and "join and leave" rather than a constant pool.